### PR TITLE
[Reshape] Refactor kernel functions

### LINKED ISSUE #448: Kernel refactor for `Reshape`

### PR DESCRIPTION
**Pull request by virena**
**Description of changes:**

This PR removes kernel functions from `cast.h` and `cast.cc` into separate `cast_kernels.h`, `cast_kernels.cu`, and `cast_kernels.cpp` files.

**Related Issues:**

Linked Issues:
- Issue #303 

Issues closed by this PR:
- Closes #448

**Before merging:**

- [x] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules? **n/a**

### CODE REVIEW

File: include/flexflow/ops/kernels/reshape_kernels.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+public:
+  ReshapeMeta(FFHandler handler);
+  DataType data_type;
+};
+
+namespace Kernels {
+namespace Reshape {
+
+template <typename T>
+void forward_kernel_wrapper(const T *input_ptr,
```
Comment (lockshaw) [inline include/flexflow/ops/kernels/reshape_kernels.h:20]:
```suggestion
void forward_kernel_wrapper(T const *input_ptr,
```

File: src/ops/kernels/reshape_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+#include <hip/hip_runtime.h>
+
+namespace FlexFlow {
+
+ReshapeMeta::ReshapeMeta(FFHandler handler) : OpMeta(handler) {}
+
+namespace Kernels {
+namespace Reshape {
+
+/*static*/
```
Comment (lockshaw) [inline src/ops/kernels/reshape_kernels.cpp:27]:
```suggestion
```

File: src/ops/kernels/reshape_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                                       hipStream_t stream);
+template void backward_kernel<int64_t>(int64_t *input_grad_ptr,
+                                       int64_t const *output_grad_ptr,
+                                       size_t num_elements,
+                                       hipStream_t stream);
+
+} // namespace Internal
+} // namespace Reshape
+} // namespace Kernels
+}; // namespace FlexFlow
```
Comment (lockshaw) [inline src/ops/kernels/reshape_kernels.cpp:143]: why does one have a semicolon and the others don't?

File: src/ops/kernels/reshape_kernels.cu [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+#include "flexflow/utils/cuda_helper.h"
+
+namespace FlexFlow {
+
+ReshapeMeta::ReshapeMeta(FFHandler handler) : OpMeta(handler) {}
+
+namespace Kernels {
+namespace Reshape {
+
+/*static*/
```
Comment (lockshaw) [inline src/ops/kernels/reshape_kernels.cu:26]:
```suggestion
```

File: src/ops/kernels/reshape_kernels.cu [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                                       cudaStream_t stream);
+template void backward_kernel<int64_t>(int64_t *input_grad_ptr,
+                                       int64_t const *output_grad_ptr,
+                                       size_t num_elements,
+                                       cudaStream_t stream);
+
+} // namespace Internal
+} // namespace Reshape
+} // namespace Kernels
+}; // namespace FlexFlow
```
Comment (lockshaw) [inline src/ops/kernels/reshape_kernels.cu:155]: semicolon?

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: src/ops/kernels/reshape_kernels.cpp (hunk 3 above)
Comment (virena) [reply src/ops/kernels/reshape_kernels.cpp:143]: Either is fine, I'll remove it for consistency.

**Review by lockshaw — APPROVED** [review]

**Pull request merged**